### PR TITLE
CASSANDRA-19158

### PR DESCRIPTION
**Pull request by ifesdjeen**

### CODE REVIEW

File: src/java/org/apache/cassandra/tcm/EpochAwareDebounce.java [hunk 1]
```diff
@@ -62,66 +45,46 @@ private EpochAwareDebounce()
...
-                    }
-                    catch (Throwable t)
-                    {
-                        fetchResult.cancel(true);
-                        inflightRequests.remove(fetchResult);
-                        promise.setFailure(t);
-                    }
-                });
-                return promise;
+                EpochAwareFuture promise = new EpochAwareFuture(epoch, fetchFunction.get());;
```
Comment (beobal) [inline src/java/org/apache/cassandra/tcm/EpochAwareDebounce.java:71]: nit: `;;`

File: src/java/org/apache/cassandra/tcm/EpochAwareDebounce.java [hunk 2]
```diff
@@ -62,66 +45,46 @@ private EpochAwareDebounce()
      * calls to peers, retrieving a LogState which can be applied locally to produce the necessary {@code
      * ClusterMetadata}. The function takes a {@code Promise<LogState>} as input, with the expectation that this
      * specific instance will be used to provide blocking behaviour when making the rpc calls that fetch the {@code
```
Comment (beobal) [inline src/java/org/apache/cassandra/tcm/EpochAwareDebounce.java:48]: should remove this too: `The function takes a {@code Promise<LogState>} as input, with the expectation that this specific instance will be used to provide blocking behaviour when making the rpc calls that fetch the {@code
LogState}.`

File: src/java/org/apache/cassandra/tcm/EpochAwareDebounce.java [hunk 3]
```diff
@@ -62,66 +45,46 @@ private EpochAwareDebounce()
...
-     * LogState}. These promises are memoized in order to cancel them when {@link #shutdownAndWait(long, TimeUnit)} is
-     * called. This causes the fetch function to stop waiting on any in flight {@code LogState} requests and prevents
-     * shutdown from being blocked.
+     * LogState}.
      *
      * @param  fetchFunction executes the request for LogState. It's expected that this popluates fetchResult with the
```
Comment (beobal) [inline src/java/org/apache/cassandra/tcm/EpochAwareDebounce.java:51]: And probably reword this too

File: src/java/org/apache/cassandra/tcm/PeerLogFetcher.java [hunk 4]
```diff
@@ -74,40 +74,54 @@ public ClusterMetadata fetchLogEntriesAndWait(InetAddressAndPort remote, Epoch a
...
+                log.append(logState);
+                ClusterMetadata fetched = log.waitForHighestConsecutive();
+                if (fetched.epoch.isEqualOrAfter(awaitAtleast))
+                {
+                    TCMMetrics.instance.peerLogEntriesFetched(before, logState.latestEpoch());
+                    return fetched;
+                }
+                else
+                {
+                    throw new IllegalStateException(String.format("Queried for epoch %s, but could not catch up", awaitAtleast));
```
Comment (beobal) [inline src/java/org/apache/cassandra/tcm/PeerLogFetcher.java:111]: do we want to throw here, or let the caller decide how to handle a failure to catch up? In some cases, this is purely a background task, so perhaps we should log but allow the caller to proceed? (I know we are certain that `remote` should be able to satisfy the request as we call this in response to seeing `awaitAtLeast` in some message from it).

File: src/java/org/apache/cassandra/tcm/RemoteProcessor.java [hunk 5]
```diff
@@ -152,34 +150,37 @@ public ClusterMetadata fetchLogAndWait(Epoch waitFor, Retry.Deadline retryPolicy
 
     public static ClusterMetadata fetchLogAndWait(CandidateIterator candidateIterator, LocalLog log)
     {
-        Promise<LogState> remoteRequest = new AsyncPromise<>();
-        return fetchLogAndWaitInternal(remoteRequest, candidateIterator, log);
+        try
+        {
+            return fetchLogAndWaitInternal(candidateIterator, log).awaitUninterruptibly().get();
```
Comment (beobal) [inline src/java/org/apache/cassandra/tcm/RemoteProcessor.java:155]: Should we `awaitThrowUncheckedOnInterrupt`, or even just `await` here? I'm thinking of tests really where the origin implementation (before CASSANDRA-19514) could cause occasional timeouts.

**Review by beobal — CHANGES_REQUESTED** [review]
LGTM, just a handful of minor comments. There are some test failures in relevant places too that we should investigate.

File: src/java/org/apache/cassandra/tcm/PeerLogFetcher.java (hunk 4 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/tcm/PeerLogFetcher.java:111]: Right now, we will log this (rare) exception, which ji think is a reasonable behaviour. 

On the other hand, if we have a listener, it will have to deal with exception one way or the other: this seems ok, too.

File: src/java/org/apache/cassandra/tcm/RemoteProcessor.java (hunk 5 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/tcm/RemoteProcessor.java:155]: Do you mean `await` with a timeout, or without? My logic for doing it this way was that we will rely on messaging service timeouts now to time this one out, so it should never be indefinite.

File: src/java/org/apache/cassandra/tcm/PeerLogFetcher.java (hunk 4 above)
Comment (beobal) [reply src/java/org/apache/cassandra/tcm/PeerLogFetcher.java:111]: ok, seems reasonable

File: src/java/org/apache/cassandra/tcm/RemoteProcessor.java (hunk 5 above)
Comment (beobal) [reply src/java/org/apache/cassandra/tcm/RemoteProcessor.java:155]: fair enough, I was concerned about in-flight requests preventing a timely shutdown, but this will now be bounded by `new Retry.Backoff(TCMMetrics.instance.fetchLogRetries));`, right?
